R/bookmark-state.R

```diff
@@ -0,0 +1,960 @@
+        return(NULL)
+
+      # Mark this name as pending. Use unique so that it's not added twice.
+      private$pending <- unique(c(private$pending, name))
+      private$values[[name]]
+    },
+
+    # Take pending names and mark them as used, then clear pending list.
+    flushPending = function() {
+      private$used <- unique(c(private$used, private$pending))
+      private$pending <- character(0)
+    },
+
+    asList = function() {
+      as.list.environment(private$values)
+    }
+  )
+)
+
+
+restoreCtxStack <- Stack$new()
+
+withRestoreContext <- function(ctx, expr) {
+  restoreCtxStack$push(ctx)
+
+  on.exit({
+    # Mark pending names as used
+    restoreCtxStack$peek()$flushPending()
+    restoreCtxStack$pop()
+  }, add = TRUE)
+
+  force(expr)
+}
+
+# Is there a current restore context?
+hasCurrentRestoreContext <- function() {
+  restoreCtxStack$size() > 0
+}
+
+# Call to access the current restore context
+getCurrentRestoreContext <- function() {
+  ctx <- restoreCtxStack$peek()
+  if (is.null(ctx)) {
+    stop("No restore context found")
+  }
+  ctx
+}
+
+#' Restore an input value
+#'
+#' This restores an input value from the current restore context. It should be
+#' called early on inside of input functions (like \code{\link{textInput}}).
+#'
+#' @param id Name of the input value to restore.
+#' @param default A default value to use, if there's no value to restore.
+#'
+#' @export
+restoreInput <- function(id, default) {
+  # Need to evaluate `default` in case it contains reactives like input$x. If we
+  # don't, then the calling code won't take a reactive dependency on input$x
+  # when restoring a value.
+  force(default)
+
+  if (!hasCurrentRestoreContext()) {
+    return(default)
+  }
+
+  oldInputs <- getCurrentRestoreContext()$input
+  if (oldInputs$available(id)) {
+    oldInputs$get(id)
+  } else {
+    default
+  }
+}
+
+#' Update URL in browser's location bar
+#'
+#' This function updates the URL in the client browser's location bar. It
+#' typically is called from an observer.
+#'
+#' @param queryString The new query string to show in the location bar.
+#' @param session A Shiny session object.
+#' @export
+updateLocationBar <- function(queryString, session = getDefaultReactiveDomain()) {
+  session$updateLocationBar(queryString)
+}
+
+#' Create a button for bookmarking/sharing
+#'
+#' A \code{bookmarkButton} is a \code{\link{actionButton}} with a default label
+#' that consists of a link icon and the text "Share...". It is meant to be used
+#' for bookmarking state.
+#'
+#' @param title A tooltip that is shown when the mouse cursor hovers over the
+#'   button.
+#'
+#' @seealso configureBookmarking
+#' @inheritParams actionButton
+#' @export
+bookmarkButton <- function(label = "Bookmark...",
+  icon = shiny::icon("link", lib = "glyphicon"),
+  title = "Bookmark this application's state and get a URL for sharing.",
+  ...)
+{
+  actionButton("__bookmark__", label, icon, title = title, ...)
```

Comment:
I don't suppose you've tested this with `submitButton`? I have a feeling the id needs to start with `.`.